### PR TITLE
Search specific indexes rather than wildcard query

### PR DESCRIPTION
**Pull request by roborourke**
This improves the performance on large datasets by only querying relevant indexes but also adding request caching where appropriate to improve look ups on indexes with static data e.g. those older than one day.

### CODE REVIEW

**Comment by roborourke** [comment]
cc @kevinlangleyjr - after cache kicked in I found between a 2,000% - 10,000% performance increase for some queries like audiences with this approach

File: inc/utils/namespace.php [hunk 1]
```diff
@@ -155,8 +194,49 @@ function query( array $query, array $params = [], string $path = '_search', stri
...
+			break;
+		}
+	}
+
 	// Get URL.
-	$url = add_query_arg( $params, get_elasticsearch_url() . '/analytics*/' . $path );
+	$url = add_query_arg( $params, sprintf(
+		'%s/%s/%s',
+		get_elasticsearch_url(),
+		implode( ',', $index_paths ),
```
Comment (shadyvb) [inline inc/utils/namespace.php:237]: Should we add any guards against too many selected indices here ?

**Review by shadyvb — CHANGES_REQUESTED** [review]

File: inc/utils/namespace.php [hunk 2]
```diff
@@ -142,6 +142,45 @@ function get_elasticsearch_version() : ?string {
...
+			'Analytics: Could not fetch analytics indexes: %s',
+			// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+			$indices_response->get_error_message()
+		), E_USER_WARNING );
+		return [];
+	}
+
+	if ( wp_remote_retrieve_response_code( $indices_response ) !== 200 ) {
+		trigger_error( sprintf(
+			"Analytics: ElasticSearch index deletion failed:\n%s",
```
[the remark this answers is not in the document]
Comment (roborourke) [reply inc/utils/namespace.php:169]: 🤦‍♂️  good catch

File: inc/utils/namespace.php (hunk 1 above)
Comment (roborourke) [reply inc/utils/namespace.php:237]: Yeah maybe. Elasticsearch allows up to 4kb in the URL - with 90 analytics indexes it's 1800 ascii characters (20 characters per index name) plus 89 commas.

Comment (roborourke) [reply inc/utils/namespace.php:237]: I'm not sure it's worth it looking at it, we cap the number of indexes at 90 and you'd get an error reported and a `null` value returned.

Comment (shadyvb) [reply inc/utils/namespace.php:237]: If it exceeds the cap we can query for the wildcard instead, otherwise we'd still have the error/null return.

Comment (roborourke) [reply inc/utils/namespace.php:237]: True, I've updated it for that. I suppose if someone really really wanted to they could unschedule the index cleanup routine.

**Review by shadyvb — APPROVED** [review]

**Comment by roborourke** [comment]
Thanks @shadyvb

**Pull request merged**